### ISSUE TITLE
Various ways the kernel is tripped because of changing prompts

### ISSUE THREAD

**Issue by mcaceresb** [kylebarron/stata_kernel#48]
#### Code Sample

```stata
disp "Sup?", _request(what_is_up)

pause on
program foo
    pause
    disp "bar"
end
foo

`local'

$global
```
#### Problem description

Each of the 4 examples above messes up the kernel because the prompt changes.

#### Package Version

```
Name: stata-kernel
Version: 1.1.0
Summary: A Jupyter kernel for Stata. Works with Windows, macOS, and Linux. Preserves program state.
Home-page: https://github.com/kylebarron/stata_kernel
Author: Kyle Barron
Author-email: barronk@mit.edu
License: GPLv3
Location: /usr/lib/python3.6/site-packages
Requires: jupyter-client, IPython, ipykernel, python-dateutil, pexpect
Required-by: 
```

**Comment by kylebarron** [comment]
Can you try the `dev` branch?

**Comment by mcaceresb** [comment]
Cool beans. It still gets tripped, tho. Btw is the code structure in `dev` stable? I'm meaning to give mata support another go.

**Comment by kylebarron** [comment]
Actually ``` `local'``` and `$global` might be hard to fix. Because there's no way for me to know that it's invalid input.

**Comment by mcaceresb** [comment]
Can you give me an example where using

```
re.compile(r'\r\n\. \Z', flags=re.MULTILINE)
```

to detect the end of Stata output fails? At a glance that would fix:

```
`local'
$global
qui do file
```

Then `\s\. \Z` seems to capture the first two cases; it can print something like "interactive output is not supported" or "incomplete input; aborting execution" or similar. The docstring notes "false positives" but I don't really know how to generate them.

**Comment by mcaceresb** [comment]
This just happened to me accidentally: `!` enters a bash prompt for me. I can technically use bash inside stata inside bash inside stata... While it's tempting to redirect "!" and "shell" through python, it won't resolve globals and locals correctly. Maybe pipe the statement through `disp` and then use `subprocess`?

**Comment by kylebarron** [comment]
`\Z`, like `$` in the regex would do nothing:
https://pexpect.readthedocs.io/en/stable/overview.html#find-the-end-of-line-cr-lf-conventions

**Comment by mcaceresb** [comment]
> Pexpect reads from the child one character at a time

I'm not entirely see how that makes sense. Since passing the compiled regex seems to work on my system, can you humor me and find an example where it fails so I can let it go as "what works for me is not cross platform"?

**Comment by mcaceresb** [comment]
Speaking of which, another one that trips it up for me:

```
mata printf("\r\n\r\n. ")
```

**Comment by kylebarron** [comment]
Well pexpect reads the input characters as a stream. It doesn't know when it's at the end of the stream. 

However I'm a little confused with testing. I don't know why pexpect doesn't stop after either of the first two `\r\n\. `. Maybe I'll have to try this on Mac as well.
![image](https://user-images.githubusercontent.com/15164633/44037702-f85876f4-9ee2-11e8-8663-3a42ca1e3c39.png)

The general idea is that the expect characters must not show up anywhere except as a prompt. If you do `di _n(2) ". "` that will mess up pexpect as well. But ideally this is the most peculiar regex least likely to show up as a result.

**Comment by kylebarron** [comment]
If interested you could see how pexpect deals with regex. I haven't looked into it closely.

https://github.com/pexpect/pexpect/blob/ab7d99a670794fc2b0365440340b899a346422da/pexpect/expect.py#L220-L306

**Comment by kylebarron** [comment]
> > Pexpect reads from the child one character at a time
>
> I'm not entirely sure how that makes sense. Since passing the compiled regex seems to work on my system, can you humor me and find an example where it fails so I can let it go as "what works for me is not cross platform"? If it's too much of a hassle let me know and I can let it go anyway.

From my understanding, it reads from the child `n` characters at a time, where `n` is the `searchwindowsize` option. If it's read `2n` characters and the regex matches, then it stops. I haven't looked into how it would use the `\Z`.

**Comment by mcaceresb** [comment]
It seems that the number of characters read at a time are `maxread` as set in `spawn`; see [here](https://github.com/pexpect/pexpect/blob/ab7d99a670794fc2b0365440340b899a346422da/pexpect/pty_spawn.py). For example, if I start stata via `spawn` with `maxread = 1` then my regex gets messed up. 

However, in that case the current regex gets messed up as well, along with the above. It honestly seems impossible to do this adequately via regex. I mean, you could cheat and append something to the end of each command and ignore it later. Send `.StataKernelDone` or something like that. Not at all satisfying.

**Comment by kylebarron** [comment]
If you set maxread to 1 the current regex gets messed up? Displaying some final text might be most helpful. 

Another thing to think about is how to make Stata stop on error inside a block. If you send a block from Jupyter Notebook, for example, that has `program test`, where `test` has already been defined, the entire program will be sent to Stata console, but the first line will error. So the following lines will be executed and not defined as a program.

**Comment by mcaceresb** [comment]
It does with the example you were using.

Instead of displaying text, I reckon it's faster to just hack it by declaring an object. Though if it's just going to be at the end of whatever is executed such minor speed bumps shouldn't matter.

You _could_ create a do file and pass the output via `do tempfile.do`; you'd have to add a hook at the end of the file to loop through all the locals and `c_local` them up to the current environment. Executing via `do` will stop execution of the file by default.

**Comment by kylebarron** [comment]
I had considered using a do file, or at least giving the option of using a do file, originally, but I thought it would lose all the locals by default. What is `c_local`?

**Comment by kylebarron** [comment]
Also the example I gave was an example of what I thought _should_ stop pexpect early. I was surprised that it continued until the actual end of the line.

**Comment by mcaceresb** [comment]
```
disp `hi'
program hi
c_local hi 1234
end
hi
disp `hi'
```

You van use the code I have for local completions to get all locals and then loop them up via `c_local`

**Comment by mcaceresb** [comment]
Does this work?

```
cap noi {
// code
}

mata : st_local("StataKernelLocals", invtokens(st_dir("local", "macro", "*")'))
foreach lvar of local StataKernelLocals {
c_local `lvar`: copy local `lvar'
}
exit _rc
```

**Comment by mcaceresb** [comment]
You should look into the oxymoronically named `help undocumented`

**Comment by kylebarron** [comment]
This hangs for me:
![image](https://user-images.githubusercontent.com/15164633/44045211-d45261f8-9ef5-11e8-84e5-c07cb3c70b05.png)

**Comment by kylebarron** [comment]
Even after reading through `help undocumented` I _still_ can't find `_c_local`.

**Comment by mcaceresb** [comment]
> This hangs for me:

You mean on the kernel? Yes, with the current regexes that would hang. You mean it hands with the experimental ones?

**Comment by kylebarron** [comment]
Yes. I didn't realize that would hang with the current regex. Is that because of the `qui`?

**Comment by mcaceresb** [comment]
Yes.

**Comment by kylebarron** [comment]
But even with `qui`, there's two newlines and a `.` at the end?
![image](https://user-images.githubusercontent.com/15164633/44045795-aa67c886-9ef7-11e8-82a0-ecde7b103eba.png)

**Comment by mcaceresb** [comment]
You had the `}` in the same line; that prevents the blank  line.

**Comment by mcaceresb** [comment]
`qui <command>` usually prints an extra blank line but not `do`

**Comment by kylebarron** [comment]
I see. I would be amenable to always running a do file if we can keep the locals in the environment. If we do ``cap noi do `our_file_with_commands'``, will it _always_ return two newlines and a dot prompt? Will that make mata support impossible?

**Comment by mcaceresb** [comment]
I have never liked the "do file" workaround. I reckon it's fine for certain blocks if there really is no way around it, but it messes up some things. For one, I'm pretty sure that settings inside a do file do not get set for the global environment.

**Comment by mcaceresb** [comment]
Delimit also gets reset.

**Comment by kylebarron** [comment]
I agree that do files are not desirable. But what settings? Locals and globals or other settings?

**Comment by mcaceresb** [comment]
I mean, everything in `set` no?

**Comment by kylebarron** [comment]
I've never used set. I already know the state of the delimiter prior and after the code chunk, so the #delimit isn't an issue. I would just add `#delimit ;` to the beginning of the code.

**Comment by mcaceresb** [comment]
I use it quite a bit (at least some of them), but I'm not sure which settings are session-wide and which are for the do file...

I mean, you wouldn't ignore `options` calls in R, right? Perhaps most of them are session-wide so it might not be a huge deal, but I really don't know.

The locals situation can't be resolved properly because a user might set a local to blank, and then it would be left unmodified in the parent environment.

**Comment by kylebarron** [comment]
I'm agreeing that it isn't an optimal solution.

Setting a local to blank isn't an issue anyways is it? Because the parent environment can always do `` `undefinedmacro'`` and it'll be the same if it isn't defined vs is defined and is blank.

Would you like to try the shorter regex for a day and see if anything messes up for you?

**Comment by mcaceresb** [comment]
I was thinking that you might get around this by just using `include` instead of `do`, but it seems that depending on how the do file ends that might not print the blank line, but it's a nicer workaround than `do`.

> Setting a local to blank isn't an issue anyways is it? Because the parent environment can always do `undefinedmacro' and it'll be the same if it isn't defined vs is defined and is blank.

```
local x = 1
do tempfile.do
```

if `test.do` contains `local x` then the user had asked `x` be set to blank, but it would be left unmodified. I think `include` is a thing to consider, tho.

**Comment by mcaceresb** [comment]
I can def use include in stata 13. Tried the server and it works in 11.

**Comment by mcaceresb** [comment]
How about

```
.
disp "Stata kernel done; please don't try to match this line the kernel will hang if you do."
```

**Comment by kylebarron** [comment]
Ok. Maybe this evening I'll test out `include`? Would this be something we'd use on all platforms? Or just console mac and Linux? Let's summarize:

Pros:
- Always get a repeatable dot prompt at the end of user code (is this true?, is there a way to make this true?)
- stops on error

Cons:
- More difficult to use with Mata? Would have to run Mata separately?

---
I don't follow that. In most real-world cases people won't be _trying_ to match a single period. They'll just have some type of output that happens to have `. ` at the beginning of the line, and that will just confuse them?

Also, in case you were interested we could try out [teletype](https://teletype.atom.io/). But then you'd have to settle for vim mode inside Atom instead of vim... :stuck_out_tongue:

**Comment by mcaceresb** [comment]
Re: Pros

- No, you don't always get a dot. You'd need to add some foo.
- Yes, it does stop on error.
- It has the parent locals available, modifies `set` and locals as expected.

Cons:
- I wrote a wrapper that sends stata(`"cmd"') if in mata. If you head the file with `mata` and end with `end' you'd be (mostly) fine.
- Mata execution does not stop on error with include OR with do.
- It's almost surely slower, and the lag for simple one-liners (pwd, cd, desc, etc.) might be annoying.

> I don't follow that. In most real-world cases people won't be trying to match a single period. 

I mean, that output would not be shown to the users. It would just be for the internals to match, so I was just being funny. To go to the opposite end of the spectrum, you could use a hash based on the code + time of execution and display that at the end (e.g. pass `<hash>' to execute), then you would wait for pexpect to match

```
`<hash>'\r\n. 
```

**Comment by kylebarron** [comment]
> To go to the opposite end of the spectrum, you could use a hash based on the code + time of execution and display that at the end (e.g. pass `' to execute), then you would wait for pexpect to match

I'd actually been seriously considering that. You wouldn't even need a hash, just a timestamp would be fine.

The issue is if there's malformed input. If someone writes
```
if 1 == 1 {
di "hi"
```
and doesn't close the brace, then ``di "`timestamp'"`` won't display anything.

The question is whether the lexer can be good enough to definitively stop malformed input from running.

Also, it's only the console that proxies `do_is_complete`. Hydrogen and I think Jupyter notebook just send the code straight to `do_execute`. So I need to add a message like `stata_kernel: Code block given was not complete; not sent to Stata`.

**Comment by kylebarron** [comment]
> It's almost surely slower, and the lag for simple one-liners (pwd, cd, desc, etc.) might be annoying

How much slower could it be? How long does it take to write a short string to a text file in Python and read it in Stata?

**Comment by mcaceresb** [comment]
> I'd actually been seriously considering that. You wouldn't even need a hash, just a timestamp would be fine.

I would be queasy about that because a timestamp is just a number... You want to hedge against P(accidentally matching string). At any rate, I think appending

```
`something unique'
```

to each command is an idea worth thinking about. I don't think using `disp` is a good idea precisely because of the issue you mentioned. Doing it the way I suggest requires something longer than 32 characters to make sure nothing gets printed. This would just get ignored by Stata in line continuations, etc. so it could be used in place of the first `\r\n` and solve all the issues we've pointed out here, I think. A SHA1 hash would give something that is a.s. unique and def longer than 32 characters.

**Comment by mcaceresb** [comment]
> How much slower could it be? How long does it take to write a short string to a text file in Python and read it in Stata?

On my system atm each call to `do` adds ~0.05s in lag. That adds up for short one liners. If it ends up being a lag of 0.5s you can _really_ tell, specially if you're used to a 0.01s delay. 0.05s is fine; 0.1 is pushing it. More than that becomes annoying.

**Comment by kylebarron** [comment]
Well we could first to a simple string search and if we find `program`, `if`, `else`, `qui`, `cap`, etc, then use the `include` method. For most inline entries, there will be no issue, and we can save the 0.1s lag for _most_ stuff. Especially because usually a block means you have several lines and might be running an entire program.

It doesn't actually have to be `display`. We see both the input and the output, so it can be `* <sha1>`. And the `*` should reset the dot prompt as well.

```
include user_code.do
* <sha1> from Python
```
![image](https://user-images.githubusercontent.com/15164633/44049947-87813698-9f03-11e8-94bd-954cfeab4906.png)

**Comment by kylebarron** [comment]
> `.` defines a class so `. * this is a comment` is an error on my system

I only gave `*`, not `. *`. The second dot is from the earlier `include` statement.

How long does creating the SHA1 hash take? MD5 is faster right?

**Comment by mcaceresb** [comment]
md5 is also 32 characters, right? That's a valid local.

An issue for me is that `*` is not a comment in mata (`.` does not work either). Why not just `SHA1' by itself? If we do child.pexpect_exact("`SHA1'") the prompt would not be a problem, right?

**Comment by kylebarron** [comment]
> SHA1 by itself

So you mean running it as a command? / Running it as a string?

**Comment by mcaceresb** [comment]
Actually, locals are up to 31 so you can do md5, I suppose. I thought it was collision-prone? I guess that won't matter for this application.

No, I mean as a local. One should be able to put something like
```
`78c99092b6fa549a0e656879b8c1be5920e06df0'
```

anywhere in code and it would not change anything, right? The only danger, maybe, is if the preceding character is `\`, but there is a newline before anyway, right?

**Comment by kylebarron** [comment]
[hash speeds in MB/s](https://stackoverflow.com/a/2723941):
```
MD5       407
SHA-1     312
SHA-256   148
SHA-512   189
```

For data in the kb range it might make a .1 ms difference? haha

Yes, we would impose a leading newline. I think that would be fine.

**Comment by mcaceresb** [comment]
I do the appending thing already for mata. I do `""` to force a double prompt and if I don't get it I assume an error. The `sha` idea is better, tho.

**Comment by kylebarron** [comment]
> Why not just SHA1

I was confused because you typed `` `SHA1'``. You have to double up on the ` to have embedded backticks:
```
`` `SHA1'``
```

**Comment by mcaceresb** [comment]
I am not the formatting guru you are.

**Comment by mcaceresb** [comment]
Actually, md5 is probably better to hedge against `set linesize 40`

**Comment by mcaceresb** [comment]
I still think this should only rely on `include` for code blocks, not individual lines.

I think the former sounds much nicer. If you are doing a single stream of output, it would be really nice to print output as it comes. This is esp useful for "long"-running jobs (not even something that would take hours, since that is not really an interactive sesh anymore, but even something that takes a minute or two).

I don't necessarily like the latter, if I understand the two correctly.

**Comment by kylebarron** [comment]
> I still think this should only rely on include for code blocks, not individual lines.

Agreed.

The two methods are different ways that both send output to the user as it comes in. I'll prototype each.

I have to get at least a few more hours of, you know, my actual job, in today...

**Comment by mcaceresb** [comment]
Every query of the medicare data is so slow, though. I have a real hard time finding good servers lately. Anyway, will check back later.

**Comment by kylebarron** [comment]
https://github.com/kylebarron/medicare_utils/
Working with parquet files is so fast and frees you from using only the sas servers.

**Comment by kylebarron** [comment]
I find it interesting that `include` prints the rc twice:
![image](https://user-images.githubusercontent.com/15164633/44058166-4b0227c8-9f1b-11e8-9633-63c1c7cbd5c4.png)

**Comment by mcaceresb** [comment]
Been dogfooding my mata branch this afternoon (bunch of local fixes). Actually got some stuff done, but the lag for even modestly sized code blocks feels like death.

On the other hand, context-aware mata completions is legit very helpful.

**Comment by kylebarron** [comment]
Where's the lag coming from?

**Comment by mcaceresb** [comment]
No idea. Parsing? Completions refresh? Haven't profiled (as I said, I was actually working)

**Comment by kylebarron** [comment]
I see. I think completions refresh could benefit from making the package async. Ideally there would be a bunch of jobs in the thread pool and the completions refresh takes a lower priority to almost anything else. So it would ideally refresh only while the user isn't running something.

**Comment by mcaceresb** [comment]
The problem is that they can press `tab` whenever and then what? It's only triggered if some flag `compeltions_done` is True?

**Comment by kylebarron** [comment]
Maybe it's not possible, but I envisioned a constant `env` store that `do_complete` would always use. Each use of `refresh` would entirely replace the current values of `env`, but any requests to `do_complete` while Stata is running would use the current state of `env`. So if Stata is running 100% and the user continually is asking for more things to be done, `do_complete` would be using the old `env` for a little while. 

Perhaps set it to automatically refresh every 5 commands or something. But _most_ commands won't entirely reset the current name environment.

**Comment by mcaceresb** [comment]
I added a %completions magic. Maybe have some options?

```
%completions [{on|off}] [--refresh {interval|continuous}] [--every 5]
```

**Comment by mcaceresb** [comment]
I made a timer magic and profiled ALL the things. Here you can see the time it takes state to actually run the code is negligible, but the time it takes for the kernel to interact with Stata is over a second. Completions were not the issue.

![magic_timer_lag](https://user-images.githubusercontent.com/4219197/44064452-feea740e-9f32-11e8-8ab6-cdd9d044a67c.png)

**Comment by mcaceresb** [comment]
(In case it's not clear, do_console is from StataSession; I made a mistake in printing)

**Comment by kylebarron** [comment]
Wow. Is this a command that prints a ton of output?

Can you change `pexpect.expect()` to [`expect_exact()`](https://pexpect.readthedocs.io/en/stable/api/pexpect.html#pexpect.spawn.expect_exact)? 

Maybe try changing the buffer size?

**Comment by mcaceresb** [comment]
It actually prints nothing. `self.child.sendline` is the culprit. It has a ~`0.05` lag on my system. Repeated calls really add up. %time actually starts counting from _after_ that line to hedge for that bias.

**Comment by mcaceresb** [comment]
I guess `async` can help with that? If Stata output can be shown interactively (if you run several commands at once and they all return a blank line, print only one blank line?) and the lag can be brought down, then I could start trying this out as a daily driver (the lag is a _real_ annoyance for me, btw, so if you want help with `async` let me know).

**Comment by kylebarron** [comment]
Is it easier for Mata if there's a separate log for each command or if there's a single log open the entire time? I'm thinking of trying to move towards a single log. That also could bring down latency a little?

**Comment by mcaceresb** [comment]
While I'm all for an IO stream, a literal log file on disk might be problematic... ATM the latency threshold is 0.05s. I'm sure that the lag from Stata executing a command to writing it on a log file is very small, but is it faster than 0.05s? If ther eis a fixed overhead in writing it might not be faster...

**Comment by kylebarron** [comment]
How can I test the time it takes for Stata to write to file?

**Comment by mcaceresb** [comment]
Not sure... Maybe something like:

```
fh = open("test.log", r)
fh.read()

child.sendline("test program")
timer = time()
while fh.read() == '':
    passs

print("Delay was {0:.3f}".format(time() - time))
```

On the plus side, that wouldn't have to pause for each chunk... Ah, trade-offs. Is there a drawback to the entire code block via `child.sendline`? Just checking for errors?

**Comment by mcaceresb** [comment]
`do` programs have `nostop` option to not pause on errors. A `%nostop` magic might be in order to have this behave as if literally copy/pasting commands to the terminal/GUI.

**Comment by kylebarron** [comment]
With `%timeit` I also see 50ms as the amount of time to run `child.sendline('di "hello world"')`. I agree that's very long.

So it's clear we want to limit the number of `child.sendline`s. Ideally no more than one.

Currently, one benefit of running one line at a time is exporting and loading the graph. So instead we should put splice `graph exports` into the code before sending as a big chunk?

**Comment by kylebarron** [comment]
I'm considering almost completely jettisoning using `pexpect.expect` and doing our own custom `expect`-like function on a continuously open log file. That would be more cross platform and would make sure any console advancements also appear on Windows.

**Comment by mcaceresb** [comment]
I'm realizing how much I like to use magics to set kernel options...Add 

```
%plot [--check {after | keywords}]
```

or something like that. So it would only check for graphs after code execution or as each line prints.The former is de-facto the way the GUI works, right? Since as graphs print, a window is brought up but the next graph replaces it.

However, if you want to have all the things then we could add the graph keywords to the lexer, `Token.Stata.Graphs` or something, and then if that token appears append a `graph export` statement.

**Comment by mcaceresb** [comment]
`pexpect.expect` seems like less of a problem than `child.sendline` tbh. I don't mind the hash-based check. I worry about the lag in using a log file. I mean, you're on Samsung's 960 evo, right? That should be comically fast already. How is the lag on a spinning disk?

**Comment by kylebarron** [comment]
Also I'm thinking of a different way to do graphs... adding
```stata
graph export graph$stata_kernel_graph_counter.filetype
global stata_kernel_graph_counter = $stata_kernel_graph_counter + 1
```
after **every** graph line, whether it's inside or outside a block. Then I use my pexpect-like `expect` function on the IOstream of the log on Windows, Mac, and Linux, looking for `file graph1.svg written in SVG format`. When I see that in Python, I load that graph and show it to the user.

I think this would work inside loops and programs as well..., no ?

**Comment by kylebarron** [comment]
Yes, but that's an expansion of abilities that people who don't know they exist don't have to use. That's slightly different than plots not working unless they use a magic.

**Comment by kylebarron** [comment]
This is also a reason to move to a single log file. https://github.com/kylebarron/stata_kernel/issues/60

**Comment by mcaceresb** [comment]
But the current code doesn't check for graphs inside programs already....

**Comment by mcaceresb** [comment]
> That's slightly different than plots not working unless they use a magic.

All I suggested is that the _last_ plot could be displayed, as opposed to _all_ the plots. _Something_ would still be shown. But as you think makes more sense.

**Comment by kylebarron** [comment]
I see. I think adding in this code is the only way to guarantee seeing all the plots...

**Comment by mcaceresb** [comment]
Yes. Bonus points that using explicitly named graphs will avoid the issue of collecting user-saved graphs by mistake.

**Comment by kylebarron** [comment]
:+1: 

btw I think the project has kinda bloomed in complexity, or at least I'm not used to working on a project with someone else. It feels harder to code when there's all these things I'm trying to think of. Like making sure I code in such a way that fixes all the needs of all the platforms while being fast, etc. Not complaining; you're just a faster and more experienced programmer.

**Comment by kylebarron** [comment]
Also I don't know async python, so I'm trying to learn

**Comment by mcaceresb** [comment]
Well, you'rea dealing with cross-platform crazyness, which I would have little patience for.

My idea would be to get this into a point  where it can be used as a daily driver, sans a bunch of legitimately rare edge cases (like requesting interactive input).

I think it's really close to that already. Bringing down the latency and resolving how to stream results would get it to a workable state (and if we get the mata branch working right).

At that point it would be largely better than the console. If we get some introspection or a way to explore the current dataset in general you could legitimately compete with the GUI, tbh.

**Comment by mcaceresb** [comment]
(Of course, people who click through menus will always use the Stata GUI, but that's neither here nor there...)

**Comment by kylebarron** [comment]
Yeah. This is true, (though you had the patience for cross-platform gtools). And it's my first big software project (as opposed to a data project).

**Comment by kylebarron** [comment]
oh my... I was just thinking, I wish there was pexpect for regular file streams that you didn't have to spawn in a terminal...

And guess what!
https://pexpect.readthedocs.io/en/stable/api/fdpexpect.html
> fdpexpect - use pexpect with a file descriptor
>
> This is like pexpect, but it will work with any file descriptor that you pass it. You are responsible for opening and close the file descriptor. This allows you to use Pexpect with sockets and named pipes (FIFOs).

**Comment by kylebarron** [comment]
Yes, you're right. I should try to minimize platform differences as much as possible. I think using `fdpexpect` will help with that.

I checked and it works with file descriptors on Windows as well.

**Comment by kylebarron** [comment]
Interestingly, the kernel somehow doesn't get stuck with `_request()`.
![image](https://user-images.githubusercontent.com/15164633/44312380-f8ed0300-a3c4-11e8-97a5-800e3e300240.png)

Likewise with `pause`. It doesn't give the ideal output but at least it doesn't crash.
![image](https://user-images.githubusercontent.com/15164633/44312403-318cdc80-a3c5-11e8-8bc7-810f9f76d1af.png)

**Comment by kylebarron** [comment]
All of the examples given originally now don't break the kernel, so I'm going to close this. Using the md5 was a great idea.
